libpseudodev/pseudodev.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Phoenix-RTOS
+ *
+ * pseudo devices
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/msg.h>
+#include <sys/file.h>
+#include "pseudodev.h"
+
+
+static struct {
+	oid_t oid;
+	unsigned int randSeed;
+} pseudo_common;
+
+
+static int pseudo_open(int id, int flags)
+{
+	(void)flags;
+
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idFull:
+		case pseudo_idRandom:
+			return EOK;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static int pseudo_close(int id)
+{
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idFull:
+		case pseudo_idRandom:
+			return EOK;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static ssize_t pseudo_read(int id, void *buf, size_t count)
+{
+	switch (id) {
+		case pseudo_idNull:
+			return (ssize_t)0;
+
+		case pseudo_idZero:
+		case pseudo_idFull:
+			memset(buf, 0, count);
+			return (ssize_t)count;
+
+		case pseudo_idRandom:
+			for (size_t ofs = 0; ofs < count; ofs += sizeof(pseudo_common.randSeed)) {
+				rand_r(&pseudo_common.randSeed);
+				memcpy((unsigned char *)buf + ofs, &pseudo_common.randSeed,
+					((count - ofs) > sizeof(pseudo_common.randSeed)) ? sizeof(pseudo_common.randSeed) : count - ofs);
+			}
+			return (ssize_t)count;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static ssize_t pseudo_write(int id, const void *buf, size_t count)
+{
+	(void)buf;
+
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idRandom:
+			return (ssize_t)count;
+
+		case pseudo_idFull:
+			return -ENOSPC;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static int pseudo_getattr(int id, int type, long long *attr)
+{
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idFull:
+		case pseudo_idRandom:
+			switch (type) {
+				case atUid:
+				case atGid:
+				case atSize:
+				case atBlocks:
+				case atMTime:
+				case atATime:
+					*attr = 0;
+					break;
+
+				case atMode:
+					*attr = 0666;
+					break;
+
+				case atType:
+					*attr = otDev;
+					break;
+				case atPort:
+					*attr = pseudo_common.oid.port;
+					break;
+
+				default:
+					return -EINVAL;
+			}
+			return EOK;
+
+		default:
+			break;
+	}
+
+	return ENODEV;
+}
+
+
+void pseudo_handleMsg(msg_t *msg, id_t id)
+{
+	id -= pseudo_common.oid.id;
+	switch (msg->type) {
+		case mtOpen:
+			msg->o.io.err = pseudo_open(id, msg->i.openclose.flags);
+			break;
+
+		case mtClose:
+			msg->o.io.err = pseudo_close(id);
+			break;
+
+		case mtRead:
+			msg->o.io.err = pseudo_read(id, msg->o.data, msg->o.size);
+			break;
+
+		case mtWrite:
+			msg->o.io.err = pseudo_write(id, msg->i.data, msg->i.size);
+			break;
+
+		case mtGetAttr:
+			msg->o.attr.err = pseudo_getattr(id, msg->i.attr.type, &msg->o.attr.val);
+			break;
+
+		case mtSetAttr:
+			msg->o.attr.err = -EINVAL;
+			break;
```

Comment:
These stuff shouldn't be needed - stat() uses file oid, not device one. On the other hand, mtLink and mtUnlink handlers are needed to avoid this issue again: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/718

Reply:
Removed. Now for all unsupported message types the multidrv is resonsible.